Drawing widget

henpe071 opened the pull request:
Added a widget that shows all drawings on the current video project. The drawing are store so all drawings on a frame are collected under their frame number which is a parent in the tree.

-The widget is dockable.
-The frame numbers are sorted.
-Clicking a frame number in the tree will jump to that frame.
-Clicking a drawing will jump to the correct frame and hightlight the drawing with a dashed line rectangle.
-Right clicking an item will open a context menu where you can choose rename or remove.
-Removing the frame number will delete all children and deleting the last drawing under a frame will delete the frame as well.
-Renaming the text item will change the text on the overlay.
-Clicking an item will automaticaly hightlight the hand tool
-With the hand tool you can move drawings with left click, edit with right click and change the line thickness with the scroll wheel.
-The text will change size with the scroll wheel.
-Current item can only be selected from the widget.
-Removed the undo/redo functionality

henpe071 committed:
Created a drawing widget

Added a new dockable, drawingwidget.
This widget has a list of all the frames that have drawings on them in the current video project.
By clicking a frame the video will jump to it.
By clicking a drawing the video will jump to it and the drawing will be set to current drawing.
Context menu with rename and remove exist. By removing a drawing it gets deleted and removeing a frame will remove that frame from the list and delete all drawings on the frame.

henpe071 committed:
Save and rename drawings

Now works on 4 draw types, rectangle, circle, line & arrow
Saving works
Rename works on the drawings, removes from the frames since they are constant.
If the last drawing on a frame are remove the frame is removed as well.
removed the undo/redo functionality (buttons are currently still there)
The drawings are now Shapeitems onstead of treeitems.

henpe071 committed:
Update text and drag/drop

The context menu for text have an option to change the text.
Enabled drag for the drawings and drop only in frame items.

TODO
not drag drawing to another frame
change order in vector
Fix crash when u have more than 3 drawings and restart the program 3 or more times, yeah

henpe071 committed:
Change edit shape

Changed how edit shape works
Added a delete shortcut for deleting the currently marked object in the drawing widget

henpe071 committed:
Fixed some crashes

The shape's names now updates correctly, which mean it no longer should say "unknown shape" after changing video.
Fixed a crash when having too many drawings (or something like that). It now iterate through the shapes in a vecor correctly.
Added a clear on the drawing widget when closing a video project.
Started changing how adding a new text works, not done.

henpe071 committed:
Remove the text rework

wait with it till the connect tool, since need to connect more signals from the drawing widget. Gonna need to connect signals back froms drawing widget to the overlay to show current item.

henpe071 committed:
Fix scroll wheel crash

Fixed crash when using scroll wheel while  no current drawing was selected.

henpe071 committed:
Updated edit shape

The point to edit is now selected upon the first mouse click and not based on the closeds point. To edit the other point just release the mouse and click close to the other point.

henpe071 committed:
Added change text size

You can change the text size by either drawing it lika a normal drawing or by using the scroll wheel. This means that there is currently no way to change the thickness of a text drawing.

henpe071 committed:
Update the draw end of the text

Updates the draw end coordinate to match the size of the text.

henpe071 committed:
Added the pen item and removed the ability to select items from the frame

Drawings can now only be selected from the drawing widget. This since it wasn't necessary and removed some problems.
Also added the pen item so it can be used in the drawing widget.

henpe071 committed:
Changed current drawing selection

Current drawing selection is now a dashed line around the drawing instead of inverting the color. The selection now neither is saved or shows up on the bookmarks.

henpe071 committed:
Text on frame and in widget are now mirrored

When updating the text in the widget the text on the frame are now updated. Works both with the rename option and doubleclicking on the item.

by removing the signal blocking in the add item to frame the save item data might not be needed.

henpe071 committed:
Cleaned up the code

fixed a virtual function in ShapeItem to get cleaner switches in drawingwidget

henpe071 committed:
Moved names to the cildren of shapes

It's now the children how have the names and they are stores as a Qstring.

All things renaming the drawings in the drawing widget should work.

henpe071 committed:
Added destructors

Destructors
Cleaned the current drawing rect after the item is deleted.

henpe071 committed:
Current item

Current item are now set in the drawing widget when loading a new overlay

henpe071 committed:
Added some comments

Comments in drawingwidget

henpe071 committed:
Merge branch 'development' into Drawing-widget

henpe071 committed:
Fixed a crash with show/hide drawings

fredrikiselius commented on ViAn/GUI/drawingwidget.cpp:
```diff
@@ -0,0 +1,361 @@
+        disconnect(m_overlay, SIGNAL(new_drawing(Shapes*, int)), this, SLOT(add_drawing(Shapes*, int)));
+        m_overlay = nullptr;
+    }
+    clear();
```
The clear function only removes gui items, not underlying data. Haven't check whether this is done properly or not yet. If not it should be

If that is the intended use ofc

henpe071 replied:
I know and this is what i looked at earlier this week in the "remove-items" branch. Right now if I put a delete in the shapeItem's destructor deleting the shape it will complain. But it's not perfect and there are still some mem leaks somewhere.

henpe071 replied:
The data are eventually deleted so it's not all leaked but it's not done optimally. the fix should be in the "remove" pull request which aims to fix a lot of memory leaks and removings items and data.

fredrikiselius commented on ViAn/GUI/drawingwidget.cpp:
```diff
@@ -0,0 +1,361 @@
+ * @param f_item
+ */
+void DrawingWidget::add_drawings_to_frame(FrameItem* f_item) {
+    std::vector<Shapes*> list = m_overlay->get_overlays()[f_item->get_frame()];
```
This type of indexing `m_overlay->get_overlays()[f_item->get_frame()]` is dangerous. If the frame number does not exist in the overlay map, the program will probably throw an exception and crash. 

Not entierly sure though, but its worth checking

fredrikiselius commented on ViAn/GUI/drawingwidget.cpp:
```diff
@@ -0,0 +1,361 @@
+ */
+void DrawingWidget::item_changed(QTreeWidgetItem* item) {
+    auto a_item = dynamic_cast<ShapeItem*>(item);
+    a_item->rename();
```
This is quite confusing as it is easy to assume that this will rename the actual tree item. But instead it actually renames the data item. Change the name of the function to increase readability. This goes for all drawing items

fredrikiselius commented on ViAn/GUI/drawingwidget.cpp:
```diff
@@ -0,0 +1,361 @@
+ * Calls upon standard dropEvent for TreeItems.
+ * @param event
+ */
+void DrawingWidget::dropEvent(QDropEvent *event) {
```
Remove?

fredrikiselius commented on ViAn/GUI/mainwindow.cpp:
```diff
@@ -519,12 +537,24 @@ void MainWindow::redo() {
-void MainWindow::delete_drawing() {
-    video_wgt->set_delete_drawing();
+void MainWindow::delete_current_drawing() {
+    video_wgt->set_delete_drawing(nullptr);
```
??

henpe071 replied:
they both call the same function which does a check (shape != nullptr) to know if it should delete current or the shape. And since mainwindow where delete_current_drawing is from don't know the current drawing i kinda had to do something like this somewhere. Though this might not be the best way.

fredrikiselius reviewed (commented):
* If a selectable drawing is marked, it is not deselected when a pen drawing is marked
* Undo/redo has no functionality?
* It is possible to step between drawings in the drawing widget using the key arrows. The selection is thus changed but the marked drawing in the video view is constant. It should changes in accordance with the widget.
* Drawings are created on the pressed event, thus making it possible to create non-existing drawings. Should probably be created on the release event instead.
* The dashed rectangle should have a black non-dashed rectangle underneath to create better contrast
* Would be useful if the drawing widget scrolled to the `FrameItem ` with the current frame number. It might also be good if all other items are closed and the current expanded. It gets rather messy when many frames have drawings on them.
* Moving drawings does not trigger the save prompt if the program is closed (neither does renaming the item in the drawing widget list)
* Not possible to deselect a drawing from the drawing widget

More comments below:

henpe071 reviewed (commented):
--The pen have a small bug that should be fixed but my thought was that the pen drawings would not be selected like the others, since the pen drawings doesn't have the start/end point the same way as the others have. And therefore the dashed rectangle would not capsule the drawing correctly.
--Undo/redo functionality is removed since it doesnt work with the widget. The buttons are still there however but they should be removed.
--The keyboard shortcuts is one thing i forgot to do but I should do that in a new branch
--What do you mean with non-existing drawings? Yes they will be pixel-big if you just click but they should be easy to edit or remove from the widget.
--Yes
--Yes
--Will look into.
--Can probably change to if you click in the widget but not on an item it will deselect the current

henpe071 committed:
Renamed rename and removed undo/redo

henpe071 committed:
Oops

Remove delete which shouldn't have been there.

henpe071 committed:
Bug fixes

Fix save prompt and a crash with delete action and deselect drawing in the widget by clicking the frame item.

henpe071 committed:
Merge branch 'development' into Drawing-widget

henpe071 committed:
Fix a merge miss

henpe071 committed:
Drawings are movable and savable

Drawings with the pen are now saved as a vector of points instead of pairs of points.

TODO the old shell when they were pairs are still here. remove? removeing means old pen drawings will not appear

henpe071 commented:
Fixed some comments and made the pen drawings selectable and dragable. To do this i changed so they are save as points instead of pairs of points.

henpe071 committed:
Fixed delete crash

and adding a new drawing will set it as current

henpe071 committed:
Removed focus on the shortcut
